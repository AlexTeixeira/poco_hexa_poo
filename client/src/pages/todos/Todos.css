.container {
    max-width: 600px;
    margin: 0 auto;
}

.column {
    display: flex;
    flex-direction: column;
}

.col{
    flex: 1;
}

.row {
    display: flex;
    flex-direction: row;
}


