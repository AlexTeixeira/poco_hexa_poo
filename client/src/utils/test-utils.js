"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderWithProviders = void 0;
const react_1 = __importDefault(require("react"));
const react_2 = require("@testing-library/react");
const react_redux_1 = require("react-redux");
const store_1 = require("core/src/application/states/app/store");
function renderWithProviders(ui, _a = {}) {
    var { preloadedState = {}, 
    // Automatically create a store instance if no store was passed in
    store = (0, store_1.setupStore)(preloadedState) } = _a, renderOptions = __rest(_a, ["preloadedState", "store"]);
    function Wrapper({ children }) {
        return react_1.default.createElement(react_1.default.StrictMode, null,
            react_1.default.createElement(react_redux_1.Provider, { store: store }, children));
    }
    // Return an object with the store and all of RTL's query functions
    return Object.assign({ store }, (0, react_2.render)(ui, Object.assign({ wrapper: Wrapper }, renderOptions)));
}
exports.renderWithProviders = renderWithProviders;
//# sourceMappingURL=test-utils.js.map