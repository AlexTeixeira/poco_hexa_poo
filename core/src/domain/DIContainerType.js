"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
let DIContainerType = {
    TodoRepository: Symbol("TodoRepository"),
    ErrorDriver: Symbol("ErrorDriver"),
};
exports.default = DIContainerType;
//# sourceMappingURL=DIContainerType.js.map