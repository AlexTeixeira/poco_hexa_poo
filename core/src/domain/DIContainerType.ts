let DIContainerType =
    {
        TodoRepository: Symbol("TodoRepository"),
        ErrorDriver: Symbol("ErrorDriver"),
    }
export default DIContainerType;


