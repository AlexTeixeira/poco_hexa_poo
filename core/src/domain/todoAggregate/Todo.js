"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Todo = void 0;
const TodoStatus_1 = require("./TodoStatus");
const ValidTodoSpecification_1 = require("./ValidTodoSpecification");
class Todo {
    constructor(id, title, description) {
        this.id = id;
        this.title = title;
        this.description = description;
        this._state = TodoStatus_1.TodoStatus.New;
        new ValidTodoSpecification_1.ValidTodoSpecification(this).isSatisfiedBy();
    }
    get state() {
        return this._state;
    }
    set state(state) {
        this._state = state;
    }
}
exports.Todo = Todo;
//# sourceMappingURL=Todo.js.map