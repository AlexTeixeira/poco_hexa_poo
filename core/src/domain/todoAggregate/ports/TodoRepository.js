"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//# sourceMappingURL=TodoRepository.js.map