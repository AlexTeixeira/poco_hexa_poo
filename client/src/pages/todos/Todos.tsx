import React, {FC, useEffect, useState} from "react";
import {useAppDispatch, useAppSelector} from "core/src/application/states/app/hooks";
import {
    selectErrorByKey,
    selectTodos,
    selectTodosByState,
    TodosState
} from "core/src/application/states/features/todo/todosSlice";
import {createTodoAsync} from "core/src/application/states/features/todo/useCases/createTodo";
import {TodoDto} from "core/src/application/states/features/todo/models/TodoDto";
import {RootState} from "core/src/application/states/app/store";
import {loadTodosAsync} from "core/src/application/states/features/todo/useCases/loadTodos";
import {TodoCard} from "./TodoCard";
import {TodoStatus} from "core/src/domain/todoAggregate/TodoStatus";
import {useSelector} from "react-redux";
import "./Todos.css";

let fakeId = 0;

export const Todos: FC = () => {
    const dispatch = useAppDispatch();

    const todoStatus = useAppSelector(state => state.todos.status)

    useEffect(() => {
        if (todoStatus === 'idle') {
            dispatch(loadTodosAsync())
        }
    }, [todoStatus])

    const useErrorByKey = (key: string) => {
        const error =
            useAppSelector((state: RootState) => selectErrorByKey(state, key));
        return <span>{error}</span>
    }

    const useSelectTodoByState = (todoState: TodoStatus) => {
        return useSelector((state: RootState) => selectTodosByState(state, todoState)) ?? [];
    }

    const [title, setTitle] = useState("");
    const [description, setDescription] = useState("");

    function addTodo() {
        fakeId = fakeId + 1;
        return () => dispatch(createTodoAsync({
            id: fakeId.toString(),
            title: title,
            description: description
        } as TodoDto));
    }

    return (
        <section className={"container"}>
            <div className={"column"}>
                <div className={"col, column"}>
                    <input className={"col"} type={"text"} placeholder={"todo.title"}
                           onInput={(event: React.ChangeEvent<HTMLInputElement>) => setTitle(event.target.value)}/>
                    <div>{useErrorByKey("title")}</div>
                    <input className={"col"} type={"text"} placeholder={"todo.description"}
                           onInput={(event: React.ChangeEvent<HTMLInputElement>) => setDescription(event.target.value)}/>
                    <div>{useErrorByKey("description")}</div>
                    <button className={"btn btn-primary"}
                            onClick={addTodo()}>todo.add
                    </button>
                </div>
                <div className={"row"}>
                    <div data-testid={"todo.card.new"}>
                        {useSelectTodoByState(TodoStatus.New).map(todo => <TodoCard key={todo.id} todo={todo}/>)}</div>
                    <div data-testid={"todo.card.inProgress"}>
                        {useSelectTodoByState(TodoStatus.InProgress).map(todo => <TodoCard key={todo.id}
                                                                                           todo={todo}/>)}</div>
                    <div data-testid={"todo.card.done"}>
                        {useSelectTodoByState(TodoStatus.Done).map(todo => <TodoCard key={todo.id} todo={todo}/>)}</div>
                </div>
            </div>
        </section>
    );
}