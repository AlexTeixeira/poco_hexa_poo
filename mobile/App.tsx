import {Button, StyleSheet, Text, TextInput, View} from 'react-native';
import {RootState,} from "core/src/application/states/app/store";
import * as React from 'react';
import {TodoStatus} from "core/src/domain/todoAggregate/TodoStatus";
import {TodoDto} from "core/src/application/states/features/todo/models/TodoDto";
import {useAppDispatch, useAppSelector} from "core/src/application/states/app/hooks";
import {selectErrorByKey, selectTodosByState,} from "core/src/application/states/features/todo/todosSlice";
import {useEffect, useState} from "react";
import {createTodoAsync} from "core/src/application/states/features/todo/useCases/createTodo";
import {useSelector} from "react-redux";
import {deleteTodoAsync} from "core/src/application/states/features/todo/useCases/deleteTodo";
import {loadTodosAsync} from "core/src/application/states/features/todo/useCases/loadTodos";

let fakeId = 0;
export default function App() {
    const dispatch = useAppDispatch();

    const useSelectTodoByState = (todoState: TodoStatus) => {
        return useSelector((state: RootState) => selectTodosByState(state, todoState)) ?? [];
    }

    const [title, setTitle] = useState("");
    const [description, setDescription] = useState("");

    const todoStatus = useAppSelector(state => state.todos.status)
    useEffect(() => {
        if (todoStatus === 'idle') {
            dispatch(loadTodosAsync())
        }
    }, [todoStatus])

    function addTodo() {
        fakeId = fakeId + 1;
        dispatch(createTodoAsync({
            id: fakeId.toString(),
            title: title,
            description: description
        } as TodoDto));
    }

    const useErrorByKey = (key: string) => {
        const error =
            useAppSelector((state: RootState) => selectErrorByKey(state, key));
        return <Text>{error}</Text>
    }

    return (
        <>
            <View>
                <TextInput placeholder={"todo.title"} onChangeText={(val) => setTitle(val)}></TextInput>
                {useErrorByKey("title")}
                <TextInput placeholder={"todo.description"} onChangeText={(val) => setDescription(val)}></TextInput>
                {useErrorByKey("description")}
                <Button title={"todo.add"} onPress={addTodo}></Button>
            </View>
            <View testID={"todo.card.new"}>
                {useSelectTodoByState(TodoStatus.New).map(todo => (<View key={todo.id}>
                    <Text>{todo.title}</Text>
                    <Button title={"todo.delete " + todo.id} onPress={() => dispatch(deleteTodoAsync(todo.id))}></Button>
                    </View>))}
            </View>
            <View testID={"todo.card.inProgress"}>
                {useSelectTodoByState(TodoStatus.InProgress).map(todo => (<View key={todo.id}>
                    <Text>{todo.title}</Text>
                    <Button title={"todo.delete " + todo.id} onPress={() => dispatch(deleteTodoAsync(todo.id))}></Button>
                </View>))}
            </View>
            <View testID={"todo.card.done"}>
                {useSelectTodoByState(TodoStatus.Done).map(todo => (<View key={todo.id}>
                    <Text>{todo.title}</Text>
                    <Button title={"todo.delete " + todo.id} onPress={() => dispatch(deleteTodoAsync(todo.id))}></Button>
                </View>))}
            </View>
        </>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        backgroundColor: '#fff',
        alignItems: 'center',
        justifyContent: 'center',
    },
});
