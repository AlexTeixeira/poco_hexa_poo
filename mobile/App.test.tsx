import App from "./App";
import testContainer from "core/src/config/TestDIContainer";
import {TodoRepository} from "core/src/domain/todoAggregate/ports/TodoRepository";
import DIContainerType from "core/src/domain/DIContainerType";
import {Todo} from "core/src/domain/todoAggregate/Todo";
import * as React from "react";
import {renderWithProviders} from "./utils/test-utils";
import {TodoStatus} from "core/src/domain/todoAggregate/TodoStatus";
import {act, fireEvent, screen} from "@testing-library/react-native";
import {Todos} from "client/src/pages/todos/Todos";


async function createTodosAsync(initialTodos: Todo[]) {
    const todoRepository = testContainer.get<TodoRepository>(DIContainerType.TodoRepository);
    for (let item of initialTodos) {
        await todoRepository.createAsync(new Todo(item.id, item.title, item.description))
    }
}


describe("App", () => {

    beforeEach(() => {
        const todoRepository = testContainer.get<TodoRepository>(DIContainerType.TodoRepository);
        todoRepository.clear();
    })

    const createTodoAsync = (todos: Todo[]) => {
        const todoRepository = testContainer.get<TodoRepository>(DIContainerType.TodoRepository);
        for (let todo of todos) {
            todoRepository.createAsync(todo)
        }
    }

    test('should load some todos', async () => {
        createTodoAsync([new Todo("1", "title", "description")]);
        const {findAllByTestId} = renderWithProviders(<App/>);
        const todos = await findAllByTestId("todo.card.new");
        expect(todos).toHaveLength(1);
    });

    test('should load todos in correct containers', async () => {
        createTodoAsync([
            {id: "1", title: "title 1", description: "description", state: TodoStatus.New} as Todo,
            {id: "2", title: "title 2", description: "description", state: TodoStatus.InProgress} as Todo,
            {id: "3", title: "title 3", description: "description", state: TodoStatus.Done} as Todo
        ]);
        const {findByText} = renderWithProviders(<App/>);
        expect(screen.toJSON()).toMatchSnapshot();


        expect(await findByText("title 1")).toBeDefined()
        expect(await findByText("title 2")).toBeDefined()
        expect(await findByText("title 3")).toBeDefined()
    });


    test("add a todo should display the todo", async () => {
        const {findByText, findAllByTestId, findByPlaceholderText} = renderWithProviders(<App/>);

        const titleElement = await screen.findByPlaceholderText("todo.title")
        const descriptionElement = await screen.findByPlaceholderText("todo.description")
        const buttonElement = await screen.findByText("todo.add")

        await act(() => {
            fireEvent(titleElement, "onChangeText", "title")
            fireEvent(descriptionElement, "onChangeText", "description")
        });

        await act(() => {
            fireEvent.press(buttonElement)
        });

        expect(await screen.findAllByTestId("todo.card.new")).toHaveLength(1);
        expect(await screen.findByText("title")).toBeDefined()
    });

    test("todo without title should popup an error", async () => {
        const {findByText, queryAllByTestId, findByPlaceholderText} = renderWithProviders(<App/>);

        const titleElement = await findByPlaceholderText("todo.title")
        const descriptionElement = await findByPlaceholderText("todo.description")
        const buttonElement = await findByText("todo.add")

        await act(() => {
            fireEvent(titleElement, "onChangeText", "")
            fireEvent(descriptionElement, "onChangeText", "description")
        });

        await act(() => {
            fireEvent.press(buttonElement)
        });


        await expect(await findByText("Title is required")).toBeDefined()
    });

    test("todo without description should popup an error", async () => {
        const {findByText, queryAllByTestId, findByPlaceholderText} = renderWithProviders(<App/>);

        const titleElement = await findByPlaceholderText("todo.title")
        const descriptionElement = await findByPlaceholderText("todo.description")
        const buttonElement = await findByText("todo.add")

        await act(() => {
            fireEvent(titleElement, "onChangeText", "title")
            fireEvent(descriptionElement, "onChangeText", "")
        });

        await act(() => {
            fireEvent.press(buttonElement)
        });

        expect(await findByText("Description is required")).toBeDefined()
    });

    test("delete a todo should remove the todo", async () => {

        const initialTodos: Todo[] = [{id: "741", title: "title", description: "description"} as Todo]
        await createTodosAsync(initialTodos);
        const {findByText, queryAllByTestId} = renderWithProviders(<App/>);

        const deleteButtonElement = await findByText("todo.delete 741")

        await act(() => {
            fireEvent.press(deleteButtonElement)
        });

        expect(await queryAllByTestId("todo.card")).toHaveLength(0);
    });

    // test("click on edit button should move to edit mode", async () => {
    //     const initialTodos = {
    //         status: 'idle',
    //         items: [{id: "741", title: "title", description: "description"}],
    //         errors: []
    //     }
    //     await createTodosAsync(initialTodos);
    //     const {findByText, findByDisplayValue} = renderWithProviders(<App/>);
    //
    //     const editButtonElement = await findByText("todo.edit 741")
    //     act(() => {
    //         fireEvent.click(editButtonElement)
    //     });
    //     expect(await findByDisplayValue("title")).toBeDefined()
    //     expect(await findByDisplayValue("description")).toBeDefined()
    //     expect(await findByText("todo.update")).toBeDefined()
    // });
});