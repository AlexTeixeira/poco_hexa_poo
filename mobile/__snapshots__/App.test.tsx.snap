// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App delete a todo should remove the todo 1`] = `
[
  <View>
    <TextInput
      onChangeText={[Function]}
      placeholder="todo.title"
    />
    <Text />
    <TextInput
      onChangeText={[Function]}
      placeholder="todo.description"
    />
    <Text />
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          todo.add
        </Text>
      </View>
    </View>
  </View>,
  <View
    testID="todo.card.new"
  />,
  <View
    testID="todo.card.inProgress"
  />,
  <View
    testID="todo.card.done"
  />,
]
`;

exports[`App should load todos in correct containers 1`] = `
[
  <View>
    <TextInput
      onChangeText={[Function]}
      placeholder="todo.title"
    />
    <Text />
    <TextInput
      onChangeText={[Function]}
      placeholder="todo.description"
    />
    <Text />
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          todo.add
        </Text>
      </View>
    </View>
  </View>,
  <View
    testID="todo.card.new"
  />,
  <View
    testID="todo.card.inProgress"
  />,
  <View
    testID="todo.card.done"
  />,
]
`;
